storage/redis/adapter.go

```diff
@@ -0,0 +1,101 @@
+package redis
+
+import (
+	"fmt"
+	"os"
+	"strconv"
+
+	r "gopkg.in/redis.v3"
+)
+
+type adapter struct {
+	bufferSize  int
+	redisClient *r.Client
+}
+
+// NewStorageAdapter returns a pointer to a new instance of a redis-based storage.Adapter.
+func NewStorageAdapter(bufferSize int) (*adapter, error) {
+	if bufferSize <= 0 {
+		return nil, fmt.Errorf("Invalid buffer size: %d", bufferSize)
+	}
+	db, err := getIntOpt("DEIS_LOGGER_REDIS_DB", 0)
+	if err != nil {
+		return nil, err
+	}
+	return &adapter{
+		bufferSize: bufferSize,
+		redisClient: r.NewClient(&r.Options{
+			Addr:     fmt.Sprintf("%s:%s", os.Getenv("DEIS_LOGGER_REDIS_SERVICE_HOST"), os.Getenv("DEIS_LOGGER_REDIS_SERVICE_PORT")),
+			DB:       int64(db),
+			Password: getStringOpt("DEIS_LOGGER_REDIS_PASSWORD", ""), // "" == no password
+		}),
+	}, nil
+}
+
+// Write adds a log message to to an app-specific list in redis using ring-buffer-like semantics
+func (a *adapter) Write(app string, message string) error {
+	// Note: Deliberately NOT using MULTI / transactions here since in this implementation of the
+	// redis client, MULTI is not safe for concurrent use by multiple goroutines. It's been advised
+	// by the authors of the gopkg.in/redis.v3 package to just use pipelining when possible...
+	// and here that is technically possible. In the WORST case scenario, not having transactions
+	// means we may momentarily have more than the desired number of log entries in the list /
+	// buffer, but an LTRIM will eventually correct that, bringing the list / buffer back down to
+	// its desired max size.
```

Comment:
It can happen when log entries arrive at loggers out of timestamp order. See the trivial scenario below.

1. node A receives `log1`, which is timestamped at `T`
2. node A `RPush`es `log1`
3. node B receives `log2`, which is timestamped at `T-1`
4. node B `LTrim`s & removes `log1`
5. node A `LTrim`s, but the buffer size is under the max so no entries are removed

Note: IMO it seems like this issue isn't worth dealing with right now